src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class NullableWalker
+    {
+#if DEBUG
+        /// <summary>
+        /// Verifies that all BoundExpressions in a tree have been visited by the nullable walker and have recorded updated types and nullabilities for rewriting purposes.
+        /// </summary>
+        private sealed class DebugVerifier : BoundTreeWalker
+        {
+            private readonly ImmutableDictionary<BoundExpression, (NullabilityInfo Info, TypeSymbol Type)> _analyzedNullabilityMap;
+            private readonly HashSet<BoundExpression> _visitedExpressions = new HashSet<BoundExpression>();
+            private int _recursionDepth;
+
+            private DebugVerifier(ImmutableDictionary<BoundExpression, (NullabilityInfo Info, TypeSymbol Type)> analyzedNullabilityMap)
+            {
+                _analyzedNullabilityMap = analyzedNullabilityMap;
+            }
+
+            protected override bool ConvertInsufficientExecutionStackExceptionToCancelledByStackGuardException()
+            {
+                return false; // Same behavior as NullableWalker
+            }
+
+            public static void Verify(ImmutableDictionary<BoundExpression, (NullabilityInfo Info, TypeSymbol Type)> analyzedNullabilityMap, BoundNode node)
+            {
+                var verifier = new DebugVerifier(analyzedNullabilityMap);
+                verifier.Visit(node);
+                // Can't just remove nodes from _analyzedNullabilityMap and verify no nodes remaining because nodes can be reused.
```

Comment:
>// Can't just remove nodes from _analyzedNullabilityMap and verify no nodes remaining because nodes can be reused. [](start = 16, length = 114)

The comment appears to be stale. Can it be removed?

Reply:
I suppose it's more explaining why I didn't just make _analyzedNullabilityMap mutable and remove nodes from it as I visited them, and am instead just doing a count comparison.

---
In reply to: [266118053](https://github.com/dotnet/roslyn/pull/34017#discussion_r266118053) [](ancestors = 266118053)